tensorflow/lite/kernels/stablehlo_reduce_precision.cc

```diff
@@ -0,0 +1,168 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include <cmath>
+#include <limits>
+
+#include "Eigen/Core"
+#include "tensorflow/core/platform/bfloat16.h"
+#include "tensorflow/lite/core/c/builtin_op_data.h"
+#include "tensorflow/lite/core/c/common.h"
+#include "tensorflow/lite/kernels/internal/tensor_ctypes.h"
+#include "tensorflow/lite/kernels/kernel_util.h"
+
+namespace tflite {
+namespace ops {
+namespace builtin {
+namespace stablehlo_reduce_precision {
+namespace{
+constexpr int kInputTensor = 0;
+constexpr int kOutputTensor = 0;
+
+template <typename DataType>
+struct FloatingPointTraits {
+  static constexpr int mantissa_bits =
+      std::numeric_limits<DataType>::digits - 1;
+  static constexpr int max_exponent =
+      std::numeric_limits<DataType>::max_exponent;
+  static constexpr int min_exponent =
+      std::numeric_limits<DataType>::min_exponent;
+  static constexpr int exponent_bits =
+      (__builtin_clz(static_cast<unsigned int>(max_exponent - min_exponent + 1)) ^
+       31) +
+      1;
+};
+template <typename DataType>
+DataType ReducePrecision(DataType value, int32_t exponent_bits,
+                         int32_t mantissa_bits);
+
+template <typename DataType>
+TfLiteStatus ReducePrecisionOp(const TfLiteTensor* operand,
+                               int32_t exponent_bits, int32_t mantissa_bits,
+                               TfLiteTensor* result) {
+  for (int i = 0; i < NumElements(result); ++i) {
+    DataType value = GetTensorData<DataType>(operand)[i];
+    DataType reduced_value = ReducePrecision(value, exponent_bits, mantissa_bits);
+    GetTensorData<DataType>(result)[i] = reduced_value;
+  }
+  return TfLiteStatus::kTfLiteOk;
+}
+
+template <typename DataType>
+DataType ReducePrecision(DataType value, int32_t exponent_bits,
+                         int32_t mantissa_bits) {
+  uint32_t int_val = *reinterpret_cast<uint32_t*>(&value);
+
+  const int32_t src_mantissa_bits = FloatingPointTraits<DataType>::mantissa_bits;
+  const int32_t src_exponent_bits = FloatingPointTraits<DataType>::exponent_bits;
+
+  if (mantissa_bits < src_mantissa_bits) {
+    uint32_t last_mantissa_bit_mask = 1ull << (src_mantissa_bits - mantissa_bits);
+    uint32_t base_rounding_bias = (last_mantissa_bit_mask >> 1) - 1;
+    uint32_t x_last_mantissa_bit =
+        (int_val & last_mantissa_bit_mask) >> (src_mantissa_bits - mantissa_bits);
+    uint32_t xRoundingBias = x_last_mantissa_bit + base_rounding_bias;
+
+    uint32_t truncation_mask = ~(last_mantissa_bit_mask - 1);
```

Comment:
These calculations will be identical for every call to ReducePrecision (which would be each element of the Tensor). Can we instead pre-calculate those once in ReducePrecisionOp and pass them? Same goes for the exponent calculations below.

You may want to also consider making this function either non-templated or templated based on the uint data type instead of floating point, and keeping this function limited to just the bit manipulation.

Reply:
Yes, I have shifted repeated calculations to ReducePrecisionOp and passed those values as parameters to ReducePrecision